File: Problems/0078.Subsets/Subsets.go
package leetcode

import "fmt"

// Runtime: 0 ms, faster than 100.00% of Go online submissions
func subsets(nums []int) [][]int {
	res := [][]int{{}}
	for _, num := range nums {
		for _, sub := range res {
			clone := make([]int, len(sub), len(sub)+1)
			copy(clone, sub)
			res = append(res, append(clone, num))
			// 這邊如果不做 copy ，而是用以下寫法 (直接 append 子slice 並回傳)，
			// 會有預期外的錯誤
			// res = append(res, append(sub, num))
		}
	}
	return res
}

File: Problems/0078.Subsets/Subsets.py
from typing import List
from itertools import combinations

class Solution:

    # Runtime: 28 ms, faster than 96.15% of Python3 online submissions for Subsets
    def subsets(self, nums: List[int]) -> List[List[int]]:
        return [
            list(tup)
            for i in range(len(nums)+1)
            for tup in combinations(nums, i)
        ]
    
    # Runtime: 24 ms, faster than 99.21% of Python3 online submissions for Subsets.
    def subsetsV2(self, nums: List[int]) -> List[List[int]]:
        res = [[]]
        for n in nums:
            res += [tup + [n] for tup in res ]
            print("****", res)
        return res